src/Bicep.Core/Emit/EmitLimitationCalculator.cs

```diff
@@ -93,5 +96,153 @@ public static EmitLimitationInfo Calculate(SemanticModel semanticModel)
 
             return moduleScopeData.ToImmutableDictionary();
         }
+
+        private static IEnumerable<Diagnostic> DetectDuplicateNames(SemanticModel semanticModel, ImmutableDictionary<ResourceSymbol, ResourceSymbol?> resourceScopeData, ImmutableDictionary<ModuleSymbol, ScopeHelper.ScopeData> moduleScopeData)
+        {
+            // This method only checks, if in one deployment we do not have 2 or more resources with this same name in one deployment to avoid template validation error
+            // This will not check resource constraints such as necessity of having unique virtual network names within resource group
+
+            var duplicateResources = semanticModel.Root.AllDeclarations
+                .OfType<ResourceSymbol>()
+                .Where(res => res.Type is ResourceType && res.DeclaringResource.Body is ObjectSyntax)
+                .Select(res =>
+                {
+                    var namePropSyntax = ((ObjectSyntax)res.DeclaringResource.Body).SafeGetPropertyByName(LanguageConstants.ResourceNamePropertyName);
+                    var resourceScope = resourceScopeData[res];
+                    return new
+                    {
+                        ResourceName = res.Name,
+                        ResourceScope = resourceScope,
+                        ResourceTypeFQDN = ((ResourceType)res.Type).TypeReference.FullyQualifiedType,
+                        ResourceNamePropertyValue = namePropSyntax?.Value as StringSyntax, //TODO: currently limiting check to property values that are strings, although it can be references or other syntaxes
+                        ResourceNamePropertySyntax = namePropSyntax
+                    };
+                })
+                .Where(x => x.ResourceNamePropertySyntax is not null && x.ResourceNamePropertyValue is not null)
```

Comment:
Typing this without an IDE so some of the syntax may be incorrect, but I think it would be worth:
1. Create a dedicated class for this rather than using anonymous objects, and encapsulating the equality comparer, e.g.:
    ```csharp
    private class ResourceDefinition
    {
        public ResourceSymbol Resource { get; }
        public SyntaxBase NameSyntax { get; }
        public ResourceType Type { get; }
        public ResourceSymbol? Scope { get; }

        public ResourceDefinition(ResourceSymbol resource, SyntaxBase nameSyntax, ResourceType type, ResourceSymbol? scope)
        {
            Resource = resource;
            NameSyntax = nameSyntax;
            Type = type;
            Scope = scope;
        }

        public static IEqualityComparer<ResourceDefinition> Comparer { get; } = new EqualityComparer();

        private class EqualityComparer : IEqualityComparer<ResourceDefinition>
        {
            ...
        }
    }
    ```
1. Create a generator func to return valid `ResourceDefinition` instances for a given semantic model to avoid the complex Linq expressions:
    ```csharp
    public static IEnumerable<ResourceDefinition> GetResourceDefinitions(SemanticModel model, ImmutableDictionary<ResourceSymbol, ResourceSymbol?> resourceScopeData)
    {
        foreach (var resource in model.Root.ResourceDeclarations)
        {
            var nameSyntax = resource.SafeGetBodyPropertyValue(LanguageConstants.ResourceNamePropertyName);
            var resourceType = resource.Type as ResourceType;
            if (nameSyntax is null or resourceType is null)
            {
                continue;
            }
            
            yield return new ResourceDefinition(resource, nameSyntax, resourceType, resourceScopeData[resource]);
        }
    }
    ```
1. Above logic can then be simplified to:
    ```csharp
    var duplicatedResources = GetResourceDefinitions(semanticModel)
        .GroupBy(ResourceDefinition.Comparer)
        .Where(group => group.Count() > 1);

    foreach (var duplicatedResource in duplicateResources)
    {
        ...
    }
    ```

Reply:
Refactored